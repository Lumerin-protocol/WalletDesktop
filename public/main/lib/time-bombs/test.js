const timeBombs = require('.')

const start = Date.now()

const bomb1 = timeBombs.create(250)

bomb1.onBoom(function () {
  console.log('Should show ~550ms:', Date.now() - start)
})

setTimeout(function () {
  bomb1.reset(200)
  setTimeout(function () {
    bomb1.reset(200)
  }, 150)
}, 200)

const bomb2 = timeBombs.create(100)

bomb2.onBoom(function () {
  console.log('Should show ~350ms:', Date.now() - start)
})

setTimeout(function () {
  bomb2.reset(300)
}, 50)
