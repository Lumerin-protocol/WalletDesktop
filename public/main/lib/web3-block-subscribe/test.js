const url = 'ws://parity.bloqrock.net:8546'
// const url = 'ws://localhost:8546'

const web3Subscription = require('.')

web3Subscription.subscribe({
  url,
  onData: function ({ number }) {
    console.log('Hash', number)
  },
  onError: function (err) {
    console.warn('Error', err.message)
  }
})
