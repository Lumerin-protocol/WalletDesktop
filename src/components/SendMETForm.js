import { DisplayValue, FieldBtn, TextInput, Flex, Btn, Sp } from './common'
import { sendToMainProcess, isWeiable } from '../utils'
import ConfirmationWizard from './ConfirmationWizard'
import * as validators from '../validator'
import * as selectors from '../selectors'
import { debounce } from 'lodash'
import { connect } from 'react-redux'
import GasEditor from './GasEditor'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import config from '../config'
import React from 'react'
import Web3 from 'web3'

const ConfirmationContainer = styled.div`
  font-size: 1.3rem;
  font-weight: 600;
  letter-spacing: 0.5px;

  & > div {
    color: ${p => p.theme.colors.primary};
  }
`

const Footer = styled.div`
  background-image: linear-gradient(to bottom, #272727, #323232);
  padding: 3.2rem 2.4rem;
  flex-grow: 1;
  height: 100%;
`

class SendMETForm extends React.Component {
  static propTypes = {
    availableMET: PropTypes.string.isRequired,
    from: PropTypes.string.isRequired,
    tabs: PropTypes.node
  }

  state = {
    ...GasEditor.initialState('MET'),
    toAddress: null,
    metAmount: null,
    errors: {}
  }

  onMaxClick = () => {
    const metAmount = Web3.utils.fromWei(this.props.availableMET)
    this.setState({ metAmount })
  }

  onInputChange = e => {
    const { id, value } = e.target

    this.setState(state => ({
      ...state,
      [id]: value,
      errors: { ...state.errors, [id]: null }
    }))

    // Estimate gas limit again if parameters changed
    if (['toAddress', 'metAmount'].includes(id)) this.getGasEstimate()
  }

  getGasEstimate = debounce(() => {
    const { metAmount, toAddress } = this.state

    if (!isWeiable(metAmount) || !Web3.utils.isAddress(toAddress)) return

    sendToMainProcess('tokens-get-gas-limit', {
      value: Web3.utils.toWei(metAmount.replace(',', '.')),
      token: config.MET_TOKEN_ADDR,
      from: this.props.from,
      to: toAddress
    })
      .then(({ gasLimit }) => this.setState({ gasLimit: gasLimit.toString() }))
      .catch(err => console.warn('Gas estimation failed', err)) // eslint-disable-line no-console
  }, 500)

  validate = () => {
    const { toAddress, metAmount, gasLimit, gasPrice } = this.state
    const max = Web3.utils.fromWei(this.props.availableMET)
    const errors = {
      ...validators.validateToAddress(toAddress),
      ...validators.validateMetAmount(metAmount, max),
      ...validators.validateGasPrice(gasPrice, config.MAX_GAS_PRICE),
      ...validators.validateGasLimit(gasLimit)
    }
    const hasErrors = Object.keys(errors).length > 0
    if (hasErrors) this.setState({ errors })
    return !hasErrors
  }

  renderConfirmation = () => {
    const { metAmount, toAddress } = this.state
    return (
      <ConfirmationContainer data-testid="confirmation">
        You will send{' '}
        <DisplayValue
          inline
          value={Web3.utils.toWei(metAmount.replace(',', '.'))}
          post=" MET"
        />{' '}
        to the address {toAddress}.
      </ConfirmationContainer>
    )
  }

  onWizardSubmit = password => {
    return sendToMainProcess('send-token', {
      gasPrice: Web3.utils.toWei(this.state.gasPrice.replace(',', '.'), 'gwei'),
      gasLimit: this.state.gasLimit,
      password,
      token: config.MET_TOKEN_ADDR,
      value: Web3.utils.toWei(this.state.metAmount.replace(',', '.')),
      from: this.props.from,
      to: this.state.toAddress
    })
  }

  renderForm = goToReview => {
    return (
      <Flex.Column grow="1">
        {this.props.tabs}
        <Sp py={4} px={3}>
          <form
            data-testid="sendMet-form"
            noValidate
            onSubmit={goToReview}
            id="sendForm"
          >
            <TextInput
              placeholder="e.g. 0x2345678998765434567"
              data-testid="toAddress-field"
              autoFocus
              onChange={this.onInputChange}
              error={this.state.errors.toAddress}
              label="Send to Address"
              value={this.state.toAddress}
              id="toAddress"
            />
            <Sp mt={3}>
              <FieldBtn
                data-testid="max-btn"
                tabIndex="-1"
                onClick={this.onMaxClick}
                float
              >
                MAX
              </FieldBtn>
              <TextInput
                placeholder="0.00"
                data-testid="metAmount-field"
                onChange={this.onInputChange}
                error={this.state.errors.metAmount}
                label="Amount (MET)"
                value={this.state.metAmount}
                id="metAmount"
              />
            </Sp>

            <Sp mt={3}>
              <GasEditor
                useCustomGas={this.state.useCustomGas}
                onChange={this.onInputChange}
                gasPrice={this.state.gasPrice}
                gasLimit={this.state.gasLimit}
                errors={this.state.errors}
              />
            </Sp>
          </form>
        </Sp>
        <Footer>
          <Btn block submit form="sendForm">
            Review Send
          </Btn>
        </Footer>
      </Flex.Column>
    )
  }

  render() {
    return (
      <ConfirmationWizard
        renderConfirmation={this.renderConfirmation}
        onWizardSubmit={this.onWizardSubmit}
        renderForm={this.renderForm}
        validate={this.validate}
      />
    )
  }
}

const mapStateToProps = state => ({
  availableMET: selectors.getMtnBalanceWei(state),
  from: selectors.getActiveWalletAddresses(state)[0]
})

export default connect(mapStateToProps)(SendMETForm)
