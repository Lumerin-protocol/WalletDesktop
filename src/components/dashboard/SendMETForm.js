import withSendMETFormState from 'metronome-wallet-ui-logic/src/hocs/withSendMETFormState'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import React from 'react'

import {
  ConfirmationWizard,
  DisplayValue,
  GasEditor,
  TextInput,
  FieldBtn,
  Flex,
  Btn,
  Sp
} from '../common'

const ConfirmationContainer = styled.div`
  font-size: 1.3rem;
  font-weight: 600;
  letter-spacing: 0.5px;

  & > div {
    color: ${p => p.theme.colors.primary};
  }
`

const Footer = styled.div`
  background-image: linear-gradient(to bottom, #272727, #323232);
  padding: 3.2rem 2.4rem;
  flex-grow: 1;
  height: 100%;
`

class SendMETForm extends React.Component {
  static propTypes = {
    gasEstimateError: PropTypes.bool,
    metPlaceholder: PropTypes.string,
    onInputChange: PropTypes.func.isRequired,
    useCustomGas: PropTypes.bool.isRequired,
    onMaxClick: PropTypes.func.isRequired,
    metAmount: PropTypes.string,
    toAddress: PropTypes.string,
    onSubmit: PropTypes.func.isRequired,
    validate: PropTypes.func.isRequired,
    gasPrice: PropTypes.string,
    gasLimit: PropTypes.string,
    errors: PropTypes.shape({
      metAmount: PropTypes.string,
      toAddress: PropTypes.string
    }).isRequired,
    tabs: PropTypes.node.isRequired
  }

  renderConfirmation = () => (
    <ConfirmationContainer data-testid="confirmation">
      You will send{' '}
      <DisplayValue inline value={this.props.metAmount} toWei post=" MET" /> to
      the address {this.props.toAddress}.
    </ConfirmationContainer>
  )

  renderForm = goToReview => (
    <Flex.Column grow="1">
      {this.props.tabs}
      <Sp py={4} px={3}>
        <form
          data-testid="sendMet-form"
          noValidate
          onSubmit={goToReview}
          id="sendForm"
        >
          <TextInput
            placeholder="e.g. 0x2345678998765434567"
            data-testid="toAddress-field"
            autoFocus
            onChange={this.props.onInputChange}
            error={this.props.errors.toAddress}
            label="Send to Address"
            value={this.props.toAddress}
            id="toAddress"
          />
          <Sp mt={3}>
            <FieldBtn
              data-testid="max-btn"
              tabIndex="-1"
              onClick={this.props.onMaxClick}
              float
            >
              MAX
            </FieldBtn>
            <TextInput
              placeholder={this.props.metPlaceholder}
              data-testid="metAmount-field"
              onChange={this.props.onInputChange}
              error={this.props.errors.metAmount}
              label="Amount (MET)"
              value={this.props.metAmount}
              id="metAmount"
            />
          </Sp>

          <Sp mt={3}>
            <GasEditor
              gasEstimateError={this.props.gasEstimateError}
              onInputChange={this.props.onInputChange}
              useCustomGas={this.props.useCustomGas}
              gasLimit={this.props.gasLimit}
              gasPrice={this.props.gasPrice}
              errors={this.props.errors}
            />
          </Sp>
        </form>
      </Sp>
      <Footer>
        <Btn block submit form="sendForm">
          Review Send
        </Btn>
      </Footer>
    </Flex.Column>
  )

  render() {
    return (
      <ConfirmationWizard
        renderConfirmation={this.renderConfirmation}
        onWizardSubmit={this.props.onSubmit}
        renderForm={this.renderForm}
        validate={this.props.validate}
      />
    )
  }
}

export default withSendMETFormState(SendMETForm)
