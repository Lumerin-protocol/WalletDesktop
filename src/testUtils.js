/* eslint-disable require-path-exists/exists */
/* eslint-disable import/no-unresolved */
import { Provider as ClientProvider } from 'lumerin-wallet-ui-logic/src/hocs/clientContext';
import { Provider, createStore } from 'lumerin-wallet-ui-logic/src/store';
import { render, Simulate } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import { MemoryRouter } from 'react-router';
import createClient from './client';
import { merge } from 'lodash';
import theme from 'lumerin-wallet-ui-logic/src/theme';
import React from 'react';

import config from '../config';

/*
 * The same render method of 'react-testing-library' but wrapped with
 * Redux store Provider and 'styled-components' ThemeProvider.
 *
 * element      : the React element you want to wrap
 * initialState : optional state to initialize the Redux store
 *
 * Returns the same object as render() but with an additional 'store'
 * property useful for dispatching actions inside your tests.
 */
export function reduxRender(element, initialState) {
  const client = createClient(config, createStore, initialState);

  const renderResult = render(
    <ClientProvider value={client}>
      <Provider store={client.store}>
        <ThemeProvider theme={theme}>
          <React.Fragment>{element}</React.Fragment>
        </ThemeProvider>
      </Provider>
    </ClientProvider>
  );
  return { ...renderResult, store: client.store };
}

/*
 * The same as reduxRender() above but also wrapped with 'react-router'
 * in order to test navigation flows.
 *
 * Returns the same object as reduxRender()
 */
export function routerRender(element, initialState, initialRoute = '/') {
  return reduxRender(
    <MemoryRouter initialEntries={[initialRoute]}>{element}</MemoryRouter>,
    initialState
  );
}

/*
 * Declaratively test complex form validations
 *
 * getter   : function that receives a data-testid and returns an HTML element
 * formId   : data-testid of the form element
 * formData : object of type { [data-testid]: any } to populate the form
 * errors   : object of type { [data-testid]: String } with expected errors
 */
export function testValidation(getter, formId, { formData, errors }) {
  // Populate the specified form fields
  Object.keys(formData).forEach((fieldId, i) => {
    const field = getter(fieldId);
    field.value = formData[fieldId];
    Simulate.change(field);
  });

  // Submit the form
  Simulate.submit(getter(formId));

  // Check expected errors
  Object.keys(errors).forEach((fieldId, i) => {
    expect(getter(`${fieldId}-error`).textContent).toEqual(
      expect.stringContaining(errors[fieldId])
    );
  });
}

/*
 * Add dataset property (data-foo attribute values) to nodes
 *
 * This hack is required because jsdom has no support for dataset yet
 * which prevents us from testing components that rely on it.
 * See https://github.com/jsdom/jsdom/issues/961
 *
 * element   : DOM node to add the dataset to
 * dataAttrs : attribute names to add to the node
 *
 * Usage:
 * <button data-testid="btn" data-modal="buy" data-foo="bar">Click</button>
 *
 * const btn = withDataset(getByTestId('btn'), 'modal', 'foo')
 * // btn.dataset === { modal: 'buy', foo: 'bar'}
 *
 */
export function withDataset(element, ...dataAttrs) {
  element.dataset = dataAttrs.reduce((acc, attr) => {
    acc[attr] = element.getAttribute(`data-${attr}`);
    return acc;
  }, {});
  return element;
}

/**
 * Return UTC timestamps relative to the current time
 * Useful for setting up different auction scenarios.
 */
export const twoWeeksAgo = () => Date.now() / 1000 - 60 * 60 * 24 * 7 * 2;
export const oneHourAgo = () => Date.now() / 1000 - 60 * 60;
export const inOneHour = () => Date.now() / 1000 + 60 * 60;
export const inOneWeek = () => Date.now() / 1000 + 60 * 60 * 24 * 7;

/*
 * Returns a common initial state for Redux store that is useful for most tests
 * Accepts an optional object with overrides that will be deeply merged with
 * the base state object.
 */
export function getInitialState(overrides = {}) {
  const baseState = {
    config,
    connectivity: { isOnline: true },
    blockchain: { height: 1, gasPrice: config.DEFAULT_GAS_PRICE },
    converter: {
      status: {
        availableEth: '100',
        availableLmr: '100',
        currentPrice: '10'
      }
    },
    auction: {
      status: {
        nextAuctionStartTime: inOneHour(),
        tokenRemaining: '1',
        currentAuction: 10,
        currentPrice: '33000000000',
        genesisTime: twoWeeksAgo()
      }
    },
    session: { isLoggedIn: true },
    rates: { ETH: { token: 'ETH', price: 1 } },
    wallets: {
      syncStatus: 'up-to-date',
      active: 'foo',
      allIds: ['foo'],
      byId: {
        foo: {
          addresses: {
            '0x15dd2028C976beaA6668E286b496A518F457b5Cf': {
              token: {
                [config.LMR_TOKEN_ADDR]: { balance: '5000000000000000000000' }
              },
              balance: '5000000000000000000000',
              transactions: []
            }
          }
        }
      }
    }
  };
  return merge({}, baseState, overrides);
}

/*
 * A sample transaction object. Useful for populating state in tests.
 */
export function getDummyTransaction() {
  return {
    meta: {
      lumerin: {},
      tokens: {
        [config.LMR_TOKEN_ADDR]: {
          event: 'Transfer',
          processing: false,
          value:
            '0x000000000000000000000000000000000000000000000001e5b8fa8fe2ac0000'
        }
      }
    },
    receipt: {
      blockHash: '0x1234567890abcdef',
      blockNumber: 407574,
      cumulativeGasUsed: 38484,
      gasUsed: 38484,
      transactionHash: '0x1234567890abcdef',
      transactionIndex: 0
    },
    transaction: {
      transactionIndex: 0,
      blockNumber: 407574,
      blockHash: '0x1234567890abcdef',
      hash: '0x1234567890abcdef',
      value: '0',
      gasPrice: '5000000000',
      gas: 76968,
      from: '0x15dd2028C976beaA6668E286b496A518F457b5Cf',
      to: '0xde806D6efD432CDeE42573760682D99eDEdC1d89'
    }
  };
}
