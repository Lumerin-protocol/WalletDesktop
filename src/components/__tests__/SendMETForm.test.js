import * as amountFields from './AmountFields.test.js'
import * as gasEditor from './GasEditor.test.js'
import * as testUtils from '../../testUtils'
import { Simulate } from 'react-testing-library'
import SendMETForm from '../SendMETForm'
import React from 'react'

const element = <SendMETForm />

const ETHprice = 250

const VALID_ADDRESS = '0xD6758d1907Ed647605429d40cd19C58A6d05Eb8b'

describe('<SendMETForm/>', () => {
  it('should match its snapshot', () => {
    const { container } = testUtils.reduxRender(element, getInitialState())
    expect(container).toMatchSnapshot()
  })

  describe('When editing the amount field', () => {
    it('updates MET field when MAX button is clicked', () => {
      const { getByTestId } = testUtils.reduxRender(element, getInitialState())
      const metField = getByTestId('metAmount-field')
      Simulate.click(getByTestId('max-btn'))
      expect(metField.value).toBe('5000')
    })
  })

  describe('When submitting the form', () => {
    it('displays an error if ADDRESS is not provided', () => {
      const { getByTestId } = testUtils.reduxRender(element, getInitialState())
      testUtils.testValidation(getByTestId, 'sendMet-form', {
        formData: { 'toAddress-field': '' },
        errors: { 'toAddress-field': 'Address is required' }
      })
    })

    it('displays an error if ADDRESS is invalid', () => {
      const { getByTestId } = testUtils.reduxRender(element, getInitialState())
      testUtils.testValidation(getByTestId, 'sendMet-form', {
        formData: { 'toAddress-field': 'foo' },
        errors: { 'toAddress-field': 'Invalid address' }
      })
    })

    it('displays an error if ADDRESS CHECKSUM is invalid', () => {
      const { getByTestId } = testUtils.reduxRender(element, getInitialState())
      testUtils.testValidation(getByTestId, 'sendMet-form', {
        formData: {
          'toAddress-field': '0xd6758d1907Ed647605429d40cd19C58A6d05Eb8b'
        },
        errors: { 'toAddress-field': 'Address checksum is invalid' }
      })
    })

    amountFields.runValidateTests(
      element,
      getInitialState(),
      'sendMet-form',
      'metAmount-field'
    )

    gasEditor.runValidateTests(element, getInitialState(), 'sendMet-form')

    it('displays the confirmation view if there are no errors', () => {
      const { queryByTestId, getByTestId } = testUtils.reduxRender(
        element,
        getInitialState()
      )
      expect(queryByTestId('confirmation')).toBeNull()
      const addressField = getByTestId('toAddress-field')
      addressField.value = VALID_ADDRESS
      Simulate.change(addressField)
      const amountField = getByTestId('metAmount-field')
      amountField.value = '1'
      Simulate.change(amountField)
      Simulate.submit(getByTestId('sendMet-form'))
      expect(queryByTestId('confirmation')).not.toBeNull()
    })
  })
})

function getInitialState() {
  return testUtils.getInitialState({
    rates: { ETH: { token: 'ETH', price: ETHprice } }
  })
}
