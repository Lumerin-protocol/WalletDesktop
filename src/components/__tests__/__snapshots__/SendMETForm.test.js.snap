// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AmountFields/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-jKJlTe fkNCah"
  >
    <div
      class="sc-ckVGcZ kvHIve"
    >
      <button
        type="button"
        class="sc-htpNat fpAZyb"
        data-testid="max-btn"
        tabindex="-1"
      >
        MAX
      </button>
      <div>
        <label
          class="sc-dnqmqq ilsafR"
          for="ethAmount"
        >
          Amount (ETH)
        </label>
        <input
          class="sc-iwsKbI kZiHEX"
          type="text"
          id="ethAmount"
          data-testid="ethAmount-field"
          placeholder="0.00"
          value=""
        />
      </div>
    </div>
    <div
      class="sc-hMqMXs dnTKXF"
    >
      <svg
        class="sc-bxivhb hXQRLM"
        viewBox="0 0 24 24"
      >
        <path
          d="M24 7.136v.96H1.642l4.464 4.464-.682.682-5.28-5.28a.475.475 0 0 1 0-.682L5.424 2l.682.682-4.464 4.464H24v-.01zM18.576 11.6l-.682.682 4.464 4.464H0v.96h22.358l-4.464 4.464.682.681 5.28-5.28a.475.475 0 0 0 0-.681l-5.28-5.29z"
        />
      </svg>
    </div>
    <div
      class="sc-ckVGcZ kvHIve"
    >
      <div>
        <label
          class="sc-dnqmqq ilsafR"
          for="usdAmount"
        >
          Amount (USD)
        </label>
        <input
          class="sc-iwsKbI kZiHEX"
          type="text"
          id="usdAmount"
          data-testid="usdAmount-field"
          placeholder="0.00"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<GasEditor/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-jKJlTe fkNCah"
  >
    <div
      class="sc-ckVGcZ kvHIve"
    >
      <div>
        <label
          class="sc-dnqmqq ilsafR"
          for="gasLimit"
        >
          Gas Limit (Units)
        </label>
        <input
          class="sc-iwsKbI kZiHEX"
          type="text"
          id="gasLimit"
          data-testid="gas-limit-field"
          value="21000"
        />
      </div>
    </div>
    <div
      class="sc-hMqMXs keWZLk"
    />
    <div
      class="sc-ckVGcZ kvHIve"
    >
      <div>
        <label
          class="sc-dnqmqq ilsafR"
          for="gasPrice"
        >
          Gas Price (Gwei)
        </label>
        <input
          class="sc-iwsKbI kZiHEX"
          type="text"
          id="gasPrice"
          data-testid="gas-price-field"
          value="1"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<SendMETForm/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-eNQAEJ hdurla"
  >
    <div
      class="sc-hMqMXs dKMJsu"
    >
      <form
        data-testid="sendMet-form"
        novalidate=""
        id="sendForm"
      >
        <div>
          <label
            class="sc-dnqmqq ilsafR"
            for="toAddress"
          >
            Send to Address
          </label>
          <input
            class="sc-iwsKbI kZiHEX"
            type="text"
            id="toAddress"
            placeholder="e.g. 0x2345678998765434567"
            data-testid="toAddress-field"
            value=""
          />
        </div>
        <div
          class="sc-hMqMXs bQjXee"
        >
          <button
            type="button"
            class="sc-htpNat fpAZyb"
            data-testid="max-btn"
            tabindex="-1"
          >
            MAX
          </button>
          <div>
            <label
              class="sc-dnqmqq ilsafR"
              for="metAmount"
            >
              Amount (MET)
            </label>
            <input
              class="sc-iwsKbI kZiHEX"
              type="text"
              id="metAmount"
              placeholder="0.00"
              data-testid="metAmount-field"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-hMqMXs bQjXee"
        >
          <div
            class="sc-jKJlTe kWLJKD"
          >
            <div
              class="sc-jKJlTe jGctBw"
            >
              <span
                class="sc-kEYyzF ekxbwv"
              >
                Gas Limit: 
                250000
                (Units)
              </span>
              <span
                class="sc-kEYyzF ekxbwv"
              >
                Gas Price: 
                1
                (Gwei)
              </span>
            </div>
            <button
              type="button"
              class="sc-htpNat bqwpzm"
              data-testid="edit-gas-btn"
              tabindex="-1"
            >
              EDIT GAS
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      class="sc-jDwBTQ fPkhoR"
    >
      <button
        type="submit"
        class="sc-bwzfXH gAMukr"
        form="sendForm"
      >
        Review Send
      </button>
    </div>
  </div>
</div>
`;
